File: builder/cpp/ContactBuilder.cpp
#ifndef contact_builder_cpp
#define contact_builder_cpp

#include <tuple>
#include <string>
#include <iostream>
#include "interfaces.hpp"
#include "Contact.cpp"

using namespace std;

class Contactbuilder : public ContactBuilderInterface
{
  public:
    string firstName;
    string lastName;
    string phoneNumber;

    Contactbuilder(void) {};

    void setFirstName(string name) {
      this->firstName = name;
    }

    void setLastName(string name) {
      this->lastName = name;
    }

    void setPhoneNumber(string phone) {
      this->phoneNumber = phone;
    }

    ContactInterface * build() {
      return new Contact(this->firstName + " " + this->lastName, this->phoneNumber);
    }
};

#endif

File: builder/rust/ContactBuilder.rs
use Contact::{ Contact, ContactInterface };

#[derive(Debug, Clone)]
pub struct ContactBuilder {
  first_name: String,
  last_name: String,
  phone_number: String
}

pub trait ContactBuilderInterface {
  fn new() -> Self;
  fn set_first_name(&mut self, new_name: &str) -> Self;
  fn set_last_name(&mut self, new_name: &str) -> Self;
  fn set_phone_number(&mut self, phone_number: &str) -> Self;
  fn build(&self) -> Contact;
}

impl ContactBuilderInterface for ContactBuilder {
  fn new() -> Self {
    ContactBuilder {
      first_name: "".to_string(),
      last_name: "".to_string(),
      phone_number: "".to_string()
    }
  }

  fn set_first_name(&mut self, new_name: &str) -> Self {
    self.first_name = new_name.to_string();
    self.clone()
  }

  fn set_last_name(&mut self, new_name: &str) -> Self {
    self.last_name = new_name.to_string();
    self.clone()
  }

  fn set_phone_number(&mut self, phone_number: &str) -> Self {
    self.phone_number = phone_number.to_string();
    self.clone()
  }

  fn build(&self) -> Contact {
    Contact::new(format!("{} {}", self.first_name, self.last_name), self.phone_number.clone())
  }
}